// eslint-plugin-vue/lib/rules/no-useless-template-attributes.js
/**
 * @author Yosuke Ota
 * See LICENSE file in root directory for full license.
 */
'use strict'


const utils = require('../utils')


const SPECIAL_TEMPLATE_DIRECTIVES = new Set([
  'if',
  'else',
  'else-if',
  'for',
  'slot'
])


module.exports = {
  meta: {
    type: 'problem',
    docs: {
      description: 'disallow useless attribute on `<template>`',
      categories: undefined,
      url: 'https://eslint.vuejs.org/rules/no-useless-template-attributes.html'
    },
    fixable: null,
    schema: [],
    messages: {
      unexpectedAttr: 'Unexpected useless attribute on `<template>`.',
      unexpectedDir: 'Unexpected useless directive on `<template>`.'
    }
  },
  /** @param {RuleContext} context */
  create(context) {
    /**
     * @param {VAttribute | VDirective} attr
     */
    function getKeyName(attr) {
      if (attr.directive) {
        if (attr.key.name.name !== 'bind') {
          return null
        }
        if (
          !attr.key.argument ||
          attr.key.argument.type === 'VExpressionContainer'
        ) {
          return null
        }
        return attr.key.argument.name
      }
      return attr.key.name
    }

    /**
     * @param {VAttribute | VDirective} attr
     */
    function isFragmentTemplateAttribute(attr) {
      if (attr.directive) {
        const directiveName = attr.key.name.name
        if (SPECIAL_TEMPLATE_DIRECTIVES.has(directiveName)) {
          return true
        }
        if (directiveName === 'slot-scope') {
          return true
        }
        if (directiveName === 'scope') {
          return true
        }
      }

      const keyName = getKeyName(attr)
      if (keyName === 'slot') {
        return true
      }

      return false
    }

    return utils.defineTemplateBodyVisitor(context, {
      /** @param {VStartTag} node */
      "VElement[name='template'][parent.type='VElement'] > VStartTag"(node) {
        if (!node.attributes.some(isFragmentTemplateAttribute)) {
          return
        }

        for (const attr of node.attributes) {
          if (isFragmentTemplateAttribute(attr)) {
            continue
          }
          const keyName = getKeyName(attr)
          if (keyName === 'key') {
            continue
          }
          context.report({
            node: attr,
            messageId: attr.directive ? 'unexpectedDir' : 'unexpectedAttr'
          })
        }
      }
    })
  }
}
