// eslint-plugin-node/lib/rules/no-unpublished-bin.js
/**
 * @author Toru Nagashima
 * @copyright 2016 Toru Nagashima. All rights reserved.
 * See LICENSE file in root directory for full license.
 */
"use strict"


const path = require("path")
const getConvertPath = require("../util/get-convert-path")
const getNpmignore = require("../util/get-npmignore")
const getPackageJson = require("../util/get-package-json")


/**
 * Checks whether or not a given path is a `bin` file.
 *
 * @param {string} filePath - A file path to check.
 * @param {string|object|undefined} binField - A value of the `bin` field of `package.json`.
 * @param {string} basedir - A directory path that `package.json` exists.
 * @returns {boolean} `true` if the file is a `bin` file.
 */
function isBinFile(filePath, binField, basedir) {
    if (!binField) {
        return false
    }
    if (typeof binField === "string") {
        return filePath === path.resolve(basedir, binField)
    }
    return Object.keys(binField).some(key => filePath === path.resolve(basedir, binField[key]))
}

/**
 * The definition of this rule.
 *
 * @param {RuleContext} context - The rule context to check.
 * @returns {object} The definition of this rule.
 */
function create(context) {
    return {
        Program(node) {
            let rawFilePath = context.getFilename()
            if (rawFilePath === "<input>") {
                return
            }
            rawFilePath = path.resolve(rawFilePath)

            const p = getPackageJson(rawFilePath)
            if (!p) {
                return
            }

            const basedir = path.dirname(p.filePath)
            const relativePath = getConvertPath(context)(
                path.relative(basedir, rawFilePath).replace(/\\/g, "/")
            )
            const filePath = path.join(basedir, relativePath)

            if (!isBinFile(filePath, p.bin, basedir)) {
                return
            }

            const npmignore = getNpmignore(filePath)
            if (!npmignore.match(relativePath)) {
                return
            }

            context.report({
                node,
                message:
                    "npm ignores '{{name}}'. " +
                    "Check 'files' field of 'package.json' or '.npmignore'.",
                data: {name: relativePath},
            })
        },
    }
}


module.exports = {
    create,
    meta: {
        docs: {
            description: "disallow 'bin' files that npm ignores",
            category: "Possible Errors",
            recommended: false,
        },
        fixable: false,
        schema: [
            {
                type: "object",
                properties: { //
                    convertPath: getConvertPath.schema,
                },
            },
        ],
    },
}
