// spdy/lib/spdy/response.js
'use strict'

exports.writeHead = function writeHead (statusCode, reason, obj) {
  var headers

  if (typeof reason === 'string') {
    this.statusMessage = reason
  } else {
    this.statusMessage =
      this.statusMessage || 'unknown'
    obj = reason
  }
  this.statusCode = statusCode

  if (this._headers) {
    if (obj) {
      var keys = Object.keys(obj)
      for (var i = 0; i < keys.length; i++) {
        var k = keys[i]
        if (k) this.setHeader(k, obj[k])
      }
    }
    headers = this._renderHeaders()
  } else {
    headers = obj
  }

  if (statusCode === 204 || statusCode === 304 ||
      (statusCode >= 100 && statusCode <= 199)) {
    this._hasBody = false
  }

  if (this._expect_continue && !this._sent100) {
    this.shouldKeepAlive = false
  }

  this._header = true
  this._headerSent = true

  if (this.socket._handle) { this.socket._handle._spdyState.stream.respond(this.statusCode, headers) }
}

exports.end = function end (data, encoding, callback) {
  if (!this._headerSent) {
    this.writeHead(this.statusCode)
  }

  if (!this.socket._handle) {
    return
  }

  this.finished = true

  var self = this
  var handle = this.socket._handle
  handle._spdyState.ending = true
  this.socket.end(data, encoding, function () {
    self.constructor.prototype.end.call(self, '', 'utf8', callback)
  })
}

exports.push = function push (path, headers, callback) {
  var frame = {
    path: path,
    method: headers.method ? headers.method.toString() : 'GET',
    status: headers.status ? parseInt(headers.status, 10) : 200,
    host: this._req.headers.host,
    headers: headers.request,
    response: headers.response
  }

  var stream = this.spdyStream
  return stream.pushPromise(frame, callback)
}

exports.writeContinue = function writeContinue (callback) {
  if (this.socket._handle) {
    this.socket._handle._spdyState.stream.respond(100, {}, callback)
  }
}
