// events/tests/common.js

var test = require('tape');
var assert = require('assert');

var noop = function() {};

var mustCallChecks = [];

function runCallChecks(exitCode) {
  if (exitCode !== 0) return;

  var failed = filter(mustCallChecks, function(context) {
    if ('minimum' in context) {
      context.messageSegment = 'at least ' + context.minimum;
      return context.actual < context.minimum;
    } else {
      context.messageSegment = 'exactly ' + context.exact;
      return context.actual !== context.exact;
    }
  });

  for (var i = 0; i < failed.length; i++) {
    var context = failed[i];
    console.log('Mismatched %s function calls. Expected %s, actual %d.',
        context.name,
        context.messageSegment,
        context.actual);
    if (context.stack) console.log(context.stack.split('\n').slice(2).join('\n'));
  }

  assert.strictEqual(failed.length, 0);
}

exports.mustCall = function(fn, exact) {
  return _mustCallInner(fn, exact, 'exact');
};

function _mustCallInner(fn, criteria, field) {
  if (typeof criteria == 'undefined') criteria = 1;

  if (typeof fn === 'number') {
    criteria = fn;
    fn = noop;
  } else if (fn === undefined) {
    fn = noop;
  }

  if (typeof criteria !== 'number')
    throw new TypeError('Invalid ' + field + ' value: ' + criteria);

  var context = {
    actual: 0,
    stack: (new Error()).stack,
    name: fn.name || '<anonymous>'
  };

  context[field] = criteria;

  if (mustCallChecks.length === 0) test.onFinish(function() { runCallChecks(0); });

  mustCallChecks.push(context);

  return function() {
    context.actual++;
    return fn.apply(this, arguments);
  };
}

exports.mustNotCall = function(msg) {
  return function mustNotCall() {
    assert.fail(msg || 'function should not have been called');
  };
};

function filter(arr, fn) {
  if (arr.filter) return arr.filter(fn);
  var filtered = [];
  for (var i = 0; i < arr.length; i++) {
    if (fn(arr[i], i, arr)) filtered.push(arr[i]);
  }
  return filtered
}
